module/src/hooks/clientside-validation/formValidationHook.tsx

```diff
@@ -0,0 +1,132 @@
+ * @param state the state of the form
+ * @param formProp formProp
+ */
+function runValidationOnInput<TData>(input: InputValidator, state: TData, formProp: (...keyChain: KeyChain) => BindingTools<TData>) {
+  const value = valueByKeyChain(state, input.key);
+
+  formProp(...input.key).clearValidationErrors();
+  for (const validator of input.validators) {
+    if (!validator.validate(value, state)) {
+      formProp(...input.key).addValidationError(validator.message);
+      break;
+    }
+  }
+};
+
+/**
+ * Hook to enable real time validation of an armstrong form. The form validation is defined in `inputValidators`, and
+ * the form state in `state`. This returns an object containing the following:
+ *  - `setTouched`: a method provided to tell when a form input control has been interacted with by the user. This must
+ * be used on each input with validation. As an example:
+ *
+ * ```
+ * <TextInput {...bind.text('city')} label="City" onBlur={setTouched} />
+ * ```
+ *
+ *  - `formIsValid`: a boolean flag to denote if the form is valid or invalid. This doesn't rely on user interaction, so
+ * is independant of error messages displayed on the form.
+ *  - `validationErrors`: an array populated with the errors to display on the form. Should be passed to the `DataForm` component:
+ *
+ * ```
+ * <DataForm validationResults={validationErrors} validationMode="both">
+ * ```
+ *
+ * @param inputValidators form validation
+ * @param dataBinder form state
+ */
+export function useValidatedForm<TData>(inputValidators: InputValidator[], state: TData, formProp: (...keyChain: KeyChain) => BindingTools<TData>) {
+  // Use this to calculate if this is the first render. If it is, then don't run the validation useEffect as it'll show all of the validation errors prematurely.
+  const firstRenderCheck = React.useRef(0);
+
+  // Run validation on individual inputs whenever they change value.
+  for (const inputValidator of inputValidators) {
+    React.useEffect(() => {
+      if (firstRenderCheck.current < inputValidators.length) {
+        firstRenderCheck.current++;
+        return;
+      }
+      runValidationOnInput(inputValidator, state, formProp);
+    }, [valueByKeyChain(state, inputValidator.key)]);
+  }
+
+  // Display all validation error messages on the form
+  function showAllValidation() {
+    for (const inputValidator of inputValidators) {
+      runValidationOnInput(inputValidator, state, formProp);
+    }
+  }
+
+  const formIsValid = calculateFormValidity(state, inputValidators);
+
+  // Everything to do with input state being touched below:
+  const [inputStates, setInputStates] = React.useState<InputState[]>(
+    inputValidators.map(input => ({
+      key: input.key,
+      touched: false,
+    }))
+  );
+
+  const setTouched = (inputElement: React.FocusEvent<HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement>) => {
+    const inputStatesCopy = [...inputStates];
+    const touchedInput = inputStatesCopy.find(
```

Comment:
Would this mean that the validation error comes up as soon as you start typing?

I guess if we ever did decide to pull this fully into Armstrong we'd wanna support a bunch of different options, like "dispatch on blur", "dispatch on change", "dispatch on submit" for that first option though it would need to be able to supply a blur listener to the actual inputs somehow which would potentially mean wrapping up all the inputs

Reply:
Not from this, no. 
However, validation does run on value change, but that is managed by the useEffect above.

This is essentially here to enable the user to run validation on blur. They could choose to set this to any event though, not just blur. 
The jsdocs on the hook should explain it.
```
<TextInput {...bind.text('city')} label="City" onBlur={setTouched} />
```